addons/mail/static/src/models/composer_view.js

```diff
@@ -401,6 +401,93 @@ registerModel({
+          *
+          * @param {KeyboardEvent} ev
+          */
+        onKeyupTextarea(ev) {
+            if (!this.exists()) {
+                return;
+            }
+            switch (ev.key) {
+                case 'Escape':
+                    // already handled in _onKeydownTextarea, break to avoid default
+                    break;
+                // ENTER, HOME, END, UP, DOWN, PAGE UP, PAGE DOWN, TAB: check if navigation in mention suggestions
+                case 'Enter':
+                    if (this.hasSuggestions) {
+                        this.insertSuggestion();
+                        this.closeSuggestions();
+                        this.update({ doFocus: true });
+                    }
+                    break;
+                case 'ArrowUp':
+                case 'PageUp':
+                    if (ev.key === 'ArrowUp' && !this.hasSuggestions && !this.composer.textInputContent && this.threadView) {
+                        this.threadView.startEditingLastMessageFromCurrentUser();
+                        break;
+                    }
+                    if (this.hasSuggestions) {
+                        this.setPreviousSuggestionActive();
+                        this.update({ hasToScrollToActiveSuggestion: true });
+                    }
+                    break;
+                case 'ArrowDown':
+                case 'PageDown':
+                    if (ev.key === 'ArrowDown' && !this.hasSuggestions && !this.composer.textInputContent && this.threadView) {
+                        this.threadView.startEditingLastMessageFromCurrentUser();
+                        break;
+                    }
+                    if (this.hasSuggestions) {
+                        this.setNextSuggestionActive();
+                        this.update({ hasToScrollToActiveSuggestion: true });
+                    }
+                    break;
+                case 'Home':
+                    if (this.hasSuggestions) {
+                        this.setFirstSuggestionActive();
+                        this.update({ hasToScrollToActiveSuggestion: true });
+                    }
+                    break;
+                case 'End':
+                    if (this.hasSuggestions) {
+                        this.setLastSuggestionActive();
+                        this.update({ hasToScrollToActiveSuggestion: true });
+                    }
+                    break;
+                case 'Tab':
+                    if (this.hasSuggestions) {
+                        if (ev.shiftKey) {
+                            this.setPreviousSuggestionActive();
+                            this.update({ hasToScrollToActiveSuggestion: true });
+                        } else {
+                            this.setNextSuggestionActive();
+                            this.update({ hasToScrollToActiveSuggestion: true });
+                        }
+                    }
+                    break;
+                case 'Alt':
```

Comment:
To avoid repetition, it would be possible to create a function such as `setSuggestionThenUpdate`, which does
```
function setSuggestionThenUpdate(my_function){
    if (this.hasSuggestions){
        my_function();
        this.update({ hasToScrollToActiveSuggestion: true });
    }
}
```
and call this function with the correct "setSuggestion" function as parameter.

This would avoid a lot of repetition in each switch-case

Reply:
I get the code is long and quite verbose, but I don't think this solution is good.
It would make code much more obscure with a technical shenanigan to hide important actions inside a function as parameter.

For the time being, I much prefer having everything flat.

================

Note that the code that you suggest does not scale well with adding/removing/adapting features, because it heavily couples handling of keyup events with the prompting of the "suggestion" feature.

I really think the proper way to clean things is to clearly maps keyboard shortcuts to desired action. Something like:
- just `Escape` => ...
- just `Enter` => ...
- just `Tab` => ...
- `Shift` + `Tab` => ...

It's very important to make it as flat as possible, and avoid indirection with feature bias.